robot_smach_states/src/robot_smach_states/manipulation/place.py

```diff
@@ -179,33 +186,92 @@ def execute(self, userdata):
 
 class Place(smach.StateMachine):
 
-    def __init__(self, robot, item_to_place, place_pose, arm):
+    def __init__(self, robot, item_to_place, place_pose, arm, place_volume=None, update_supporting_entity=False):
         """
         Drive the robot to be close to the designated place_pose and move the designated arm to place the designated
         item there
         :param robot: Robot to execute state with
         :param item_to_place: Designator that resolves to the entity to place. e.g EntityByIdDesignator
-        :param place_pose: Designator that resolves to the pose to place at. E.g. an EmptySpotDesignator
+        :param place_pose: The place pose can be one of three things:
+         1: Designator that resolves to the pose to place at. E.g. an EmptySpotDesignator
+         2: EdEntityDesignator resolving to an object on which the robot should place something
+         3: A string identifying an object on which the robot should place something
         :param arm: Designator -> arm to place with, so Arm that holds entity_to_place, e.g. via
         ArmHoldingEntityDesignator
+        :param place_volume (optional) string identifying the volume where to place the object, e.g., 'on_top_of',
+        'shelf3'
+        :param update_supporting_entity (optional) bool to indicate whether the supporting entity should be updated.
+        This can only be used if the supporting entity is supplied, case 2 or 3 mentioned under item_to_place
         """
         smach.StateMachine.__init__(self, outcomes=['done', 'failed'])
 
         # Check types or designator resolve types
         assert(item_to_place.resolve_type == Entity or type(item_to_place) == Entity)
-        assert(place_pose.resolve_type == FrameStamped or type(place_pose) == FrameStamped)
         assert(arm.resolve_type == Arm or type(arm) == Arm)
 
+        # Case 3
+        if isinstance(place_pose, str):
+            furniture_designator = EdEntityDesignator(robot=robot, id=place_pose)
+            place_designator = EmptySpotDesignator(robot=robot, place_location_designator=furniture_designator,
+                                                   area=place_volume)
+        # Case 1
+        elif place_pose.resolve_type == FrameStamped or type(place_pose) == FrameStamped:
```

Comment:
There is a builtin check for this in designators.checks. I think its called check_type